crates/matrix-sdk-ui/src/timeline/controller/observable_items.rs

```diff
@@ -0,0 +1,1500 @@
+    ///
+    /// Note that it doesn't clone `Self`, only the inner timeline items.
+    pub fn clone(&self) -> Vector<Arc<TimelineItem>> {
+        self.items.clone()
+    }
+
+    /// Start a new transaction to make multiple updates as one unit.
+    pub fn transaction(&mut self) -> ObservableItemsTransaction<'_> {
+        ObservableItemsTransaction {
+            items: self.items.transaction(),
+            all_remote_events: &mut self.all_remote_events,
+        }
+    }
+
+    /// Replace the timeline item at position `timeline_item_index` by
+    /// `timeline_item`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `timeline_item_index > total_number_of_timeline_items`.
+    pub fn replace(
+        &mut self,
+        timeline_item_index: usize,
+        timeline_item: Arc<TimelineItem>,
+    ) -> Arc<TimelineItem> {
+        self.items.set(timeline_item_index, timeline_item)
+    }
+
+    /// Get an iterator over all the entries in this `ObservableItems`.
+    pub fn entries(&mut self) -> ObservableItemsEntries<'_> {
+        ObservableItemsEntries(self.items.entries())
+    }
+
+    /// Call the given closure for every element in this `ObservableItems`,
+    /// with an entry struct that allows updating that element.
+    pub fn for_each<F>(&mut self, mut f: F)
+    where
+        F: FnMut(ObservableItemsEntry<'_>),
+    {
+        self.items.for_each(|entry| f(ObservableItemsEntry(entry)))
+    }
+}
+
+// It's fine to deref to an immutable reference to `Vector`.
+impl Deref for ObservableItems {
+    type Target = Vector<Arc<TimelineItem>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.items
+    }
+}
+
+/// An “iterator“ that yields entries into an `ObservableItems`.
+pub struct ObservableItemsEntries<'a>(ObservableVectorEntries<'a, Arc<TimelineItem>>);
+
+impl ObservableItemsEntries<'_> {
+    /// Advance this iterator, yielding an `ObservableItemsEntry` for the next
+    /// item in the timeline, or `None` if all items have been visited.
+    pub fn next(&mut self) -> Option<ObservableItemsEntry<'_>> {
+        self.0.next().map(ObservableItemsEntry)
+    }
+}
+
+/// A handle to a single timeline item in an `ObservableItems`.
+#[derive(Debug)]
+pub struct ObservableItemsEntry<'a>(ObservableVectorEntry<'a, Arc<TimelineItem>>);
+
+impl ObservableItemsEntry<'_> {
+    /// Replace the timeline item by `timeline_item`.
+    pub fn replace(this: &mut Self, timeline_item: Arc<TimelineItem>) -> Arc<TimelineItem> {
+        ObservableVectorEntry::set(&mut this.0, timeline_item)
+    }
+}
+
+impl Deref for ObservableItemsEntry<'_> {
+    type Target = Arc<TimelineItem>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// A transaction that allows making multiple updates to an `ObservableItems` as
+/// an atomic unit.
+///
+/// For updates from the transaction to have affect, it has to be finalized with
+/// [`Self::commit`]. If the transaction is dropped without that method being
```

Comment:
Will you ever get rid of this habit?

```suggestion
/// [`ObservableItemsTransaction::commit`]. If the transaction is dropped without that method being
```

Reply:
> Will you ever get rid of this habit?

Nah!